hub/util/diff.py

```diff
@@ -218,33 +229,84 @@ def has_change(change: Dict):
 
 
 def get_dataset_changes_for_id(
-    commit_id: str, storage: LRUCache, dataset_changes: Dict[str, Dict]
+    commit_id: str,
+    storage: LRUCache,
+    dataset_changes,
 ):
     """Returns the changes made in the dataset for a commit."""
-    if dataset_changes.get("info_updated", False):
-        # already has info_updated set to True, return
-        return
 
     dataset_diff_key = get_dataset_diff_key(commit_id)
     try:
         dataset_diff = storage.get_hub_object(dataset_diff_key, DatasetDiff)
-        dataset_changes["info_updated"] = dataset_diff.info_updated
+        dataset_changes["info_updated"] = (
+            dataset_changes.get("info_updated") or dataset_diff.info_updated
+        )
+
+        renamed = dataset_changes.get("renamed")
+        deleted = dataset_changes.get("deleted")
+        done = []
+
+        merge_renamed = OrderedDict()
+        for old, new in dataset_diff.renamed.items():
+            if deleted and new in deleted and new not in done:
+                deleted[deleted.index(new)] = old
+                done.append(old)
+                continue
+            if renamed and renamed.get(new):
+                merge_renamed[old] = renamed[new]
+                renamed.pop(new)
+            else:
+                merge_renamed[old] = new
+
+        try:
+            dataset_changes["renamed"].update(merge_renamed)
+        except KeyError:
+            dataset_changes["renamed"] = merge_renamed
+
+        if dataset_changes.get("deleted"):
+            dataset_changes["deleted"].extend(dataset_diff.deleted)
+        else:
+            dataset_changes["deleted"] = dataset_diff.deleted.copy()
     except KeyError:
         pass
 
 
 def get_tensor_changes_for_id(
-    commit_id: str, storage: LRUCache, tensor_changes: Dict[str, Dict]
+    commit_id: str,
+    storage: LRUCache,
+    tensor_changes: Dict[str, Dict],
+    dataset_changes,
 ):
     """Identifies the changes made in the given commit_id and updates them in the changes dict."""
     meta_key = get_dataset_meta_key(commit_id)
     meta = storage.get_hub_object(meta_key, DatasetMeta)
 
-    for tensor in meta.tensors:
+    for tensor, key in meta.tensor_names.items():
         try:
             commit_diff: CommitDiff
-            commit_diff_key = get_tensor_commit_diff_key(tensor, commit_id)
+            commit_diff_key = get_tensor_commit_diff_key(key, commit_id)
             commit_diff = storage.get_hub_object(commit_diff_key, CommitDiff)
+            renamed = dataset_changes.get("renamed")
+            deleted = dataset_changes.get("deleted")
+
+            if deleted and tensor in deleted:
```

Comment:
this block, till line 309 can be in a separate function

Reply:
any particular reason for this?

Reply:
we can just have a filter_deleted or some similar function that does this which simplifies and makes this function more readable.

Reply:
this does more than filtering, it continues the loop if tensor was deleted and sets which tensor the changes should be added to in line 306.